#33 Pull request to add timeout event creation

pandaadb opened the pull request:
Hi Fabien,

This is the pull request. As you know I had a bit of merge nightmare - I hope it is all readable and OK :) Let me know what you think,

Regards,
Artur

pandaadb commented:
I just signed the CLA agreement as well.

karmi commented:
Hi @pandaadb, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/logstash-plugins/logstash-filter-aggregate/pull/37.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

pandaadb commented:
@karmi hi, sorry about that. Turns out my git was misconfigured and the email address simply does not exist. Is there another way around or would I have to rebase and change my email address on every commit?

pandaadb commented:
@karmi hi again. I followed a tutorial (http://stackoverflow.com/questions/3042437/change-commit-author-at-one-specific-commit) and tried to change my Email on the commits. I think this may have doubled the commits (not really sure) but it seems to have worked? I hope that was ok. If not i think i might have to start with a clean fork and do a new pull request

fbaligand commented on CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 2.2.1
+ - new feature: Add new option "push_map_as_event_on_timeout" so that when events time out they can be yielded as new events
+ - new feature: Add new option "timeout_code" which takes the event and executes code on it. This works for "push_map_as_event_on_timeout" as well as "push_previous_map_as_event"
+
```
Given that it is clearly a new feature, the new release will be version 2.3.0.
I keep the last digit for bug fixes/minor changes.
=> Can you put 2.3.0 in changelog ?

fbaligand commented on logstash-filter-aggregate.gemspec:
```diff
@@ -1,6 +1,6 @@
   s.name = 'logstash-filter-aggregate'
-  s.version = '2.2.0'
+  s.version = '2.2.1'
   s.licenses = ['Apache License (2.0)']
```
same remark here : set version 2.3.0

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -195,6 +195,27 @@ class LogStash::Filters::Aggregate < LogStash::Filters::Base
+  # Default value: "task_id"
+  config :timeout_task_id_field, :validate => :string, :default => "task_id"
+
+
```
Up to me, no default value for this option.
So that, by default, no task_id field is set on timeout event.
It would be nice to tell that in documentation.

pandaadb commented:
Hi, 

I fixed the version numbers and changed the timeout_task_id_field default behaviour to do nothing. I updated the code + documentation to reflect this behaviour.

fbaligand commented on CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 2.3.0
+ - new feature: Add new option "push_map_as_event_on_timeout" so that when events time out they can be yielded as new events
+ - new feature: Add new option "timeout_code" which takes the event and executes code on it. This works for "push_map_as_event_on_timeout" as well as "push_previous_map_as_event"
+
```
can you also add here the feature "timeout_task_id_field" ?

pandaadb replied:
Added the option

fbaligand commented on CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 2.3.0
+ - new feature: Add new option "push_map_as_event_on_timeout" so that when events time out they can be yielded as new events
```
I think this description could be better : "so that when task timeout happens, the aggregated map can be yielded as a new event".
What do you think ?

pandaadb replied:
You're right. I updated the description

fbaligand commented on CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 2.3.0
+ - new feature: Add new option "push_map_as_event_on_timeout" so that when events time out they can be yielded as new events
+ - new feature: Add new option "timeout_code" which takes the event and executes code on it. This works for "push_map_as_event_on_timeout" as well as "push_previous_map_as_event"
```
I would replace "the event" by "the timeout event".
What you think ?

fbaligand commented on README.md:
````diff
@@ -119,24 +119,24 @@ A typical case is aggregating results from jdbc input plugin.
+            drop {}
+         }
+     }
 ```
````
If I'm not wrong, you didn't made any update in README.md ?

Up to me, this would be great to add 2 things :
- Add the new options descriptions in "Aggregate Plugin Options" section : https://github.com/logstash-plugins/logstash-filter-aggregate#aggregate-plugin-options
- Add a new example named "Example # 3" based on "push_map_as_event_on_timeout" option. And yes, that means current example # 3 becomes # 4. I think your option "push_map_as_event_on_timeout" is more generic and interest more people !

pandaadb replied:
I added an example where I count user clicks in events. The example has all 3 new options. I updated example 3 to be example 4

fbaligand replied:
Really nice example ! I love it !

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -195,6 +195,28 @@ class LogStash::Filters::Aggregate < LogStash::Filters::Base
+  # or in case 'push_previous_map_as_event' is set to true.
+  #
+  # Example value : `"map['sql_duration'] += event['duration']"`
+  config :timeout_code, :validate => :string, :required => false
```
I think we could do a better description for this option :

The code to execute to complete timeout event, when 'push_map_as_event_on_timeout' or 'push_previous_map_as_event' is set to true.  
Example value : `"event['tags'] = '_aggregatetimeout'"`

pandaadb replied:
I updates the description

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -195,6 +195,28 @@ class LogStash::Filters::Aggregate < LogStash::Filters::Base
+  # event[ "my_Id" ] = "12345"
+  #
+  config :timeout_task_id_field, :validate => :string, :required => false
+
```
I have a slight update for this option description : 
```
This option indicates the timeout generated event's field for the "task_id" value. The task id will then be set into the timeout event. This can help correlate which tasks have been timed out.  
This field has no default value and will not be set on the timeout event if not configured.
Example:
If the task_id is "12345" and this field is set to "my_Id", the generated timeout event will have :
event[ "my_Id" ] = "12345"
```

I remove "The aggregate plugin maps the aggregation map to  task_id value. ", because explaining internal plugin behaviour is not necessary here, I think. "task_id" refers to the plugin option "task_id". So I think it is not necessary to explain more about that, and this could be confusing.

pandaadb replied:
Updated the description

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -226,9 +248,15 @@ class LogStash::Filters::Aggregate < LogStash::Filters::Base
   # When this option is enabled, each time aggregate plugin detects a new task id, it pushes previous aggregate map as a new logstash event, 
   # and then creates a new empty map for the next task.
   # 
+  # When the timeout code is set, the event populated with the aggregation map is passed to the timeout code before the event is yielded
```
Up to me, this sentence is not at the right place. It is more relevant in "timeout_code".

Especially as we can use push_previous_map_as_event option without using "timeout_code" if a simple aggregation map copy is enough.

pandaadb replied:
You're right. I deleted that line.

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -226,9 +248,15 @@ class LogStash::Filters::Aggregate < LogStash::Filters::Base
+  # When this option is enabled, each time the aggreagte flushes, the aggregate map is checked for timeouts. If a timeout is detected, 
+  # the plugin will write the aggregation map into a new event. This enables storing found tasks that never receive an end task   
+  config :push_map_as_event_on_timeout, :validate => :boolean, :required => false, :default => false
   
```
I think we could make a better description, especially without explaining logstash plugin internals (like flush).
I suggest : 
When this option is enabled, each time a task timeout is detected, it pushes task aggregation map as a new logstash event.  
This enables to process task timeouts in logstash, but also to process tasks that have no explicit end event.

What you think ?

pandaadb replied:
Updated :)

fbaligand commented:
And another important thing : you can add your name in the "Contributors" section !
Right here : https://github.com/logstash-plugins/logstash-filter-aggregate/blob/master/CONTRIBUTORS

pandaadb commented:
Hey - I updated the code and added an example. As you can see I am not the best at describing things :D I hope it got more understandable now

nik-petrov commented:
Hey folks, this is great improvement! Want to put it in use asap. Is there timeline for the release?

fbaligand commented:
@nik-petrov 
I currently review the PR. As soon the review is finished, I will merge it to master and make a release.
It is soon, but I can' give you a date. Sorry.

fbaligand commented on README.md:
```diff
@@ -102,7 +102,55 @@ it allows to initialize 'sql_duration' map entry to 0 only if this map entry is
+
+A typical case is aggregating or tracking user behaviour. We can track a user by its ID through the events, however once the user stops interacting, the events stop coming in. There is no specific event indicating the end of the user's interaction.
+
+In this case, we can enable the option 'push_map_as_event_on_timeout' to enable pushing the aggregation map as a new event. 
```
At the end of this sentence, it would be nice to add "when timeout occurs".

pandaadb replied:
I added that sentence. Not sure if i got the 2 spaces right, i don't know exactly what you meant

fbaligand commented on README.md:
````diff
@@ -102,7 +102,55 @@ it allows to initialize 'sql_duration' map entry to 0 only if this map entry is
+```
+    INFO - 12345 - Clicked One
+    INFO - 12345 - Clicked Two
+    INFI - 12345 - Clicked Three
````
Replace "INFI" by "INFO"

pandaadb replied:
Whoops - corrected

fbaligand commented on CONTRIBUTORS:
```diff
@@ -6,6 +6,7 @@ Maintainers:
 
 Contributors:
 * Fabien Baligand (fbaligand)
+* Artur Kronenberg (pandaadb)
```
Nice :)

pandaadb replied:
yey :) Thank you!

fbaligand commented on README.md:
````diff
@@ -102,7 +102,55 @@ it allows to initialize 'sql_duration' map entry to 0 only if this map entry is
+In addition, we can enable 'timeout_code' to execute code on the populated timeout event.
+We can also add 'timeout_task_id_field' so we can correlate the task_id, which in this case would be the user's ID. 
+
+```
````
I think this would be nice to put an introduction point before sample, like '* So, given these logs :'
What you think ?

pandaadb replied:
Added that point

fbaligand commented on README.md:
````diff
@@ -102,7 +102,55 @@ it allows to initialize 'sql_duration' map entry to 0 only if this map entry is
+``` ruby
+    filter {
+        grok {
+                 match => [ "message", "%{LOGLEVEL:loglevel} - %{NOTSPACE:taskid} - %{GREEDYDATA:msgtext}" ]
````
to be coherent with your use case, it is maybe more relevant to name your id "userid" (instead of "taskid").
What you think ?

pandaadb replied:
Yep that makes more sense with the example

fbaligand commented on README.md:
```diff
@@ -102,7 +102,55 @@ it allows to initialize 'sql_duration' map entry to 0 only if this map entry is
+        aggregate {
+            task_id => "%{taskid}"
+            code => "map['clicks'] ||= 0; map['clicks'] += 1;"
+            timeout_task_id_field => "userId"
```
In logstash/ruby, naming convention is more either "userid", either "user_id".

That said, being a java developer (like you), I understand why you used camel case :)

pandaadb replied:
Changes msgtext to msg_text as well for consistency. Yeah, in java we love camlCaseWritings ;)

fbaligand commented on README.md:
```diff
@@ -102,7 +102,55 @@ it allows to initialize 'sql_duration' map entry to 0 only if this map entry is
+            code => "map['clicks'] ||= 0; map['clicks'] += 1;"
+            timeout_task_id_field => "userId"
+            timeout => 600 # 10 minutes timeout
+            timeout_code => "event['tags'] = '_aggregatetimeout'"
```
mmmhh, it seems that you forget the most important option : 
`push_map_as_event_on_timeout => true`

pandaadb replied:
Yep :)

pandaadb replied:
Added

fbaligand commented on README.md:
````diff
@@ -102,7 +102,55 @@ it allows to initialize 'sql_duration' map entry to 0 only if this map entry is
+
+``` json
+    {
+        "userId" : "12345",
````
same remark about naming convention

pandaadb replied:
Done

fbaligand commented on README.md:
```diff
@@ -102,7 +102,55 @@ it allows to initialize 'sql_duration' map entry to 0 only if this map entry is
+
+## Example #4
+
+Fourth use case : you have no specific start event and no specific end event.  
```
For this fourth use case, I will update documentation later to explicitly explain why it is different from your option.

pandaadb replied:
Sounds good :)

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -195,6 +195,30 @@ class LogStash::Filters::Aggregate < LogStash::Filters::Base
+  #
+  # If the taskId is "12345" and this field is set to "my_Id", the generated event will have:
+  # event[ "my_Id" ] = "12345"
+  #
```
Just a last thing about this description : 
could you replace "taskId" by "task_id" and "my_Id" by "my_id" ?

pandaadb replied:
done

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -195,6 +195,30 @@ class LogStash::Filters::Aggregate < LogStash::Filters::Base
+
+
+  # The code to execute to complete timeout generated event, when 'push_map_as_event_on_timeout' or 'push_previous_map_as_event' is set to true. 
+  # The code block will have access to the newly generated timeout event that is pre-populated with the aggregation_map values. 
```
Could you just replace "with the aggregation_map values" by "with aggregation map".
By the way, it allows to coherent this "push_map_as_event_on_timeout" description.

pandaadb replied:
Done

fbaligand commented:
When you will have fixed last little changes to the 3 new options in aggregate.rb, could you report these 3 new options documentations to README.md, section "Aggregate Plugin Options" ?

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -256,6 +283,10 @@ def register
     # process lambda expression to call in each filter call
     eval("@codeblock = lambda { |event, map| #{@code} }", binding, "(aggregate filter code)")
 
+    if @timeout_code
```
Could you just add here a comment to introduce ? 
Something like the one for @code : "process lambda expression to call in each filter call"

pandaadb replied:
Done

fbaligand commented:
Once you finalized "example #3" updates on README.md, could you report the example # 3 into aggregate.rb asciidoc documentation ?

fbaligand commented:
Endly, I start to review the code itself ! 
Sorry for long time...
Comments are coming :)

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -317,13 +348,31 @@ def filter(event)
+          # Call code block if available
+          if @timeout_code
+            begin
+              @timeout_codeblock.call(event)
```
Here, it is not "event", but "event_to_yield"

pandaadb replied:
Done

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -317,13 +348,31 @@ def filter(event)
+          if @timeout_code
+            begin
+              @timeout_codeblock.call(event)
+              events_to_flush << event
```
Up to me, in filter method, there are no events to flush.
So this line has to be removed.
Do you agree with that ?

pandaadb replied:
Yep - copy waste issue :/

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -317,13 +348,31 @@ def filter(event)
+              events_to_flush << event
+              noError = true
+            rescue => exception
+              @logger.error("Aggregate exception occurred. Error: #{exception} ; Code: #{@timeout_code} ; Map: #{map} ; EventData: #{event.instance_variable_get('@data')}")
```
To clearly separate "code" from "timeout_code" errors, could you :
- replace "Code:" by "TimeoutCode:"
- replace "Map:" by "PreviousMap:"
- replace "#{map}" by "#{previous_map}"
- replace "EventData:" by "TimeoutEventData"
- replace "event." by "event_to_yield."

pandaadb replied:
Done

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -317,13 +348,31 @@ def filter(event)
+              noError = true
+            rescue => exception
+              @logger.error("Aggregate exception occurred. Error: #{exception} ; Code: #{@timeout_code} ; Map: #{map} ; EventData: #{event.instance_variable_get('@data')}")
+              event.tag("_aggregateexception")
```
replace "event." by "event_to_yield."

pandaadb replied:
Done

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -384,10 +433,31 @@ def remove_expired_maps()
+            event = LogStash::Event.new(element.map)        
+            
+            if @timeout_task_id_field
+              event_to_yield[@timeout_task_id_field] = task_id
```
The right line should be :
`event[@timeout_task_id_field] = key`

pandaadb replied:
Good catch :)

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -384,10 +433,31 @@ def remove_expired_maps()
+            else
+              events_to_flush << event
+            end
+
```
Up to me, code `events_to_flush << event` must be present only one time just after this `end` so that generated event if always flushed, even if `timeout_code` execution failed.

pandaadb replied:
You are right. I removed the else clause and changed the code

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -384,10 +433,31 @@ def remove_expired_maps()
+              begin
+                @timeout_codeblock.call(event)
+                events_to_flush << event
+                noError = true
```
As `noError` variable is never used in this method, I suggest you to remove this line.

pandaadb replied:
Removed

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -384,10 +433,31 @@ def remove_expired_maps()
+                events_to_flush << event
+                noError = true
+              rescue => exception
+                @logger.error("Aggregate exception occurred. Error: #{exception} ; Code: #{@timeout_code} ; EventData: #{event.instance_variable_get('@data')}")
```
Could you make these updates in error log ?
- replace "Code:" by "TimeoutCode:"
- replace "EventData:" by "TimeoutEventData"

pandaadb replied:
Added this. There are also a bunch of non existent variables (i blame my editor :P) . E.g. i create "event" but I add the task_id_field to "event_to_yield". I changed all of them to be "event_to_yield" for consistency

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -384,10 +433,31 @@ def remove_expired_maps()
     min_timestamp = Time.now - @timeout
```
Can you update comment `If @push_previous_map_as_event option is set, ` to add "@push_map_as_event_on_timeout"

pandaadb replied:
Added in the comment of the remove_expired_maps method

fbaligand commented on spec/filters/aggregate_spec.rb:
```diff
@@ -10,7 +10,7 @@
     @start_filter = setup_filter({ "map_action" => "create", "code" => "map['sql_duration'] = 0" })
     @update_filter = setup_filter({ "map_action" => "update", "code" => "map['sql_duration'] += event['duration']" })
-    @end_filter = setup_filter({ "map_action" => "update", "code" => "event.to_hash.merge!(map)", "end_of_task" => true, "timeout" => 5 })
+    @end_filter = setup_filter({ "push_map_as_event_on_timeout" => true, "map_action" => "update", "code" => "event.to_hash.merge!(map)", "end_of_task" => true, "timeout" => 5, "timeout_code" => "event['test'] = 'testValue'" })
```
mmhh, doing that, you impact existing unit tests, no ?

fbaligand replied:
You should also add "timeout_task_id_field" option here, and then make some expectation about that at line 193

pandaadb replied:
Done

fbaligand commented on spec/filters/aggregate_spec.rb:
```diff
@@ -180,6 +180,20 @@
+      end
+    end
+
   end
```
Your whole new test "timeout defined on the filter" should replace previous "timeout defined on the filter" test.
It does the same checks + some others.

fbaligand replied:
Finally, it would be nice to update test "push_previous_map_as_event option is defined, " so that it verify timeout_code behaviour in `filter` method.

pandaadb replied:
Removed old test, and added some checks. Added the timeout_code test for the other test as well

fbaligand commented on spec/filters/aggregate_spec.rb:
```diff
@@ -180,6 +180,20 @@
+        sleep(2)
+        entries = @end_filter.flush()
+        expect(aggregate_maps).to be_empty
+        expect(entries.size).to eq(1)
```
You should also check here that :
- returned event entry contains at least one aggregate map element.
- returned event entry contains one field "test" with value "testValue"

pandaadb replied:
Added checks for sql_duration and testvalue

fbaligand commented on spec/filters/aggregate_spec.rb:
```diff
@@ -180,6 +180,20 @@
+        expect(aggregate_maps.size).to eq(1)
+        expect(entries).to be_empty
+      end
+      it "event is removed and creates a new event" do
```
replace by :
`it "removes event if expired and creates a new timeout event" do`

pandaadb replied:
Done

fbaligand commented:
@pandaadb 
OK, so I consider I have finished review.
Once you fixed every remark, I think I could merge your PR !

pandaadb commented:
@fbaligand I think I got them all. I'm sorry about the many mistakes, I should pay more attention to things. I added all the additional test things so they should all work now (also for the existing push previous map case). 

I am not sure if I need to document the new properties somewhere else than in the ruby file. I could not find that in the readme file. (I just saw that and am now filling that in as well)

pandaadb commented:
I added the descriptions of the new options to the readme file too. It should hopefully be all good now :)

fbaligand commented on README.md:
```diff
@@ -202,6 +252,23 @@ and then creates a new empty map for the next task.
+
+- **push_map_as_event_on_timeout**  
+When this option is enabled, each time a task timeout is detected, it pushes task aggregation map as a new logstash event.  
+This enables to detect and process task timeouts in logstash, but also to manage tasks that have no explicit end event.
```
Can you just put "push_map_as_event_on_timeout" option before "timeout_code" ?

pandaadb replied:
Yep - done :)

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -110,8 +110,61 @@
+# INFO - 12345 - SQL - sqlQuery2 - 34
+# INFO - 12345 - TASK_END - end
+# ----------------------------------
+#
```
mmmhh, not the right logs here :)

pandaadb replied:
whoops :)

fbaligand commented:
After thinking, we have a whole part that is duplicate code : timeout event creation.
And that's bad :(

So I think it would be nice to :
- add a new method called "create_timeout_event", that takes map and taskid in parameter and returns created event
- method includes aggregate.rb code from line 411 to line 427 (and equivalent code in flush method)

The only difference is line `noError = true`. And finally, this line is not relevant because the error is about previous map timeout event, and not current event that is filtered. And noError is used to know if add_field and add_tag options can be applied to current event.

What do you think about ?

pandaadb commented:
Hi Fabien,

I agree with what you said. I'll get refactoring and update my branch once it's done. It shouldn't be too big of a change

pandaadb commented:
@fbaligand I created the new method. I hope I got the coding conventions right :)

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -317,13 +401,13 @@ def filter(event)
-          previous_map = @@aggregate_maps.shift[1].map
-          event_to_yield = LogStash::Event.new(previous_map)
+          event_to_yield = create_timeout_event(@@aggregate_maps.shift[1].map, task_id)
         end
```
Can you keep line `previous_map = @@aggregate_maps.shift[1].map` ?
It allows simply to know what "@@aggregate_maps.shift[1].map" represents.

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -354,6 +438,32 @@ def filter(event)
+  # returns the newly created event
+  public 
+  def create_timeout_event(aggregation_map, task_id)
+    event_to_yield = LogStash::Event.new(aggregation_map)        
```
Can you remove "public", because `create_timeout_event` method is private to plugin ?

fbaligand commented on lib/logstash/filters/aggregate.rb:
```diff
@@ -371,23 +481,26 @@ def flush(options = {})
+
+            events_to_flush << create_timeout_event(element.map, key)
+
           end
```
OK, this is clearly not important, but can you remove blank lines 501 and 503 ?
You have right to say I am painful :)

fbaligand commented:
Ah this is so much beautiful when there is no duplicate code :)

pandaadb commented:
Hehe - your plugin, your rules :) I don't mind - I am not a ruby dev so it makes sense to do an extensive code review. I addressed your comments in the latest commit

fbaligand commented:
@pandaadb 
OK, so after a last full review, I declare that this PR is now "merge-able" !
As we say in this case : LGTM !

So I will merge it very soon (probably tonight),
And then make some documentation updates (essentially on example 4)
And then release 2.3.0 version !

Thank you a lot for this enhancement and for the time you take on this PR !

pandaadb commented:
@fbaligand it was fun :) I will see if i can write up a better use case for the second PR that I wanted to add (tracking timestamp in the event rather than system time). Next time I'll know what to look for and maybe you won't have as much review to do :) It was fun -thanks for your help!

fbaligand commented:
@nik-petrov 
release 2.3.0 is done with timeout event generation !
